From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SF2000 Linux contributors
Date: Thu, 30 Jul 2026 00:00:00 +0000
Subject: [PATCH] Respect HC15xx MIPS alignment requirements

The framebuffer is consumed by a DMA-capable presenter and must be suitably
aligned. MIPS32r1 also cannot silently repair the unaligned word accesses that
the packed emulator helper is designed to avoid.
---
 libretro/libretro.cpp   | 3 ++-
 nes_emu/blargg_common.h | 2 +-
 2 files changed, 3 insertions(+), 2 deletions(-)

diff --git a/libretro/libretro.cpp b/libretro/libretro.cpp
index b9afa9d..ad8e1a4 100644
--- a/libretro/libretro.cpp
+++ b/libretro/libretro.cpp
@@ -1330,7 +1330,8 @@ bool retro_load_game(const struct retro_game_info *info)
    static uint8_t video_buffer[Nes_Emu::image_width * (Nes_Emu::image_height + 16)];
    emu->set_pixels(video_buffer + (8 * Nes_Emu::image_width), Nes_Emu::image_width);
 #else
-   static uint8_t video_buffer[videoBufferWidth * videoBufferHeight];
+   static uint8_t video_buffer[videoBufferWidth * videoBufferHeight]
+      __attribute__((aligned(16)));
    emu->set_pixels(video_buffer, videoBufferWidth);
 #endif
 
diff --git a/nes_emu/blargg_common.h b/nes_emu/blargg_common.h
index 413fd6b..b07c196 100644
--- a/nes_emu/blargg_common.h
+++ b/nes_emu/blargg_common.h
@@ -52,7 +52,7 @@ arithmetic on smaller types. */
 #endif
 
 typedef struct
-#ifdef NO_UNALIGNED_ACCESS
+#if defined(NO_UNALIGNED_ACCESS) || defined(SF2000)
 __attribute ((packed))
 #endif
 {
-- 
2.54.0
